{{cookiecutter.repostory_name}}/app/src/{{cookiecutter.django_project_name}}/{{cookiecutter.django_default_app_name}}/consumers.py

```diff
@@ -0,0 +1,45 @@
+{%- if cookiecutter.use_channels == "y" -%}
+from collections.abc import Callable
+from typing import Annotated, ClassVar, Union
+
+import structlog
+from channels.generic.websocket import AsyncWebsocketConsumer
+from pydantic import BaseModel, Field, TypeAdapter, ValidationError
+from structlog.contextvars import bound_contextvars
+
+from .schemas import Heartbeat
+
+log = structlog.get_logger(__name__)
+
+
+class DefaultConsumer(AsyncWebsocketConsumer):
+    async def connect(self) -> None:
+        await super().connect()
+        log.info("connected", scope=self.scope)
+
+    async def disconnect(self, code: int | str) -> None:
+        log.info("disconnected", scope=self.scope, code=code)
+
+    async def receive(self, text_data: str | None = None, bytes_data: bytes | None = None) -> None:
+        with bound_contextvars(text_data=text_data, bytes_data=bytes_data):
+            log.debug("message received")
+
+            adapter = TypeAdapter(Annotated[Union[*self.MESSAGE_HANDLERS.keys()], Field(discriminator="type")])
+            try:
+                message: BaseModel = adapter.validate_json(text_data)
+            except ValidationError as exc:
+                errors = exc.errors()
+                log.debug("message parsing failed", errors=errors)
+                return
```

Comment:
I would expect something like
`self.send(text_data=Error(message=str(exc).json())`
to let the client know what happend

Then again, this is asynchronus, so perhaps this also requires that each message also has identifier of the requests so you can know for requests this is a reply for?

Reply:
Yep exactly, and the identifier is project/task-specific. May be UUID, or int, or string. So I left it for developer to decide how to implement responses.